File: src/main/python/AoC2021_09.py
#! /usr/bin/env python3
#
# Advent of Code 2021 Day 9
#

import sys
from math import prod
from typing import Iterator

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.graph import flood_fill
from aoc.grid import Cell
from aoc.grid import IntGrid

TEST = """\
2199943210
3987894921
9856789892
8767896789
9899965678
"""

Input = IntGrid
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return IntGrid.from_strings(list(input_data))

    def find_lows(self, grid: IntGrid) -> Iterator[Cell]:
        return (
            c
            for c in grid.get_cells()
            if (
                all(
                    grid.get_value(c) < grid.get_value(n)
                    for n in grid.get_capital_neighbours(c)
                )
            )
        )

    def part_1(self, grid: Input) -> int:
        return sum(grid.get_value(c) + 1 for c in self.find_lows(grid))

    def part_2(self, grid: Input) -> int:
        def size_of_basin_around_low(c: Cell) -> int:
            basin = flood_fill(
                c,
                lambda c: (
                    n
                    for n in grid.get_capital_neighbours(c)
                    if grid.get_value(n) != 9
                ),
            )
            return len(basin)

        return prod(
            sorted(
                size_of_basin_around_low(low) for low in self.find_lows(grid)
            )[-3:]
        )

    @aoc_samples(
        (
            ("part_1", TEST, 15),
            ("part_2", TEST, 1134),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2021, 9)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/julia/AoC2021_09.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2021_09

include("aoc.jl")

function _parse(input)
    grid = Matrix{Int}(undef, length(input), length(input[1]))
    for r ∈ 1:length(input), c ∈ 1:length(input[r])
        grid[CartesianIndex(r, c)] = parse(Int, input[r][c])
    end
    return grid
end

function neighbours(grid, r, c)
    return (CartesianIndex(r + Δr, c + Δc)
            for (Δr, Δc) ∈ CAPITALS
            if r + Δr ∈ 1:size(grid, 1) && c + Δc ∈ 1:size(grid, 2))
end

function find_lows(grid)
    return (CartesianIndex(r, c)
            for r ∈ 1:size(grid, 1), c ∈ 1:size(grid, 2)
            if all(grid[CartesianIndex(r, c)] < grid[n]
                   for n ∈ neighbours(grid, r, c)))
end

function part1(input)
    grid = _parse(input)
    return sum([grid[rc] + 1 for rc ∈ find_lows(grid)])
end

function part2(input)
    function sizeof_basin_round_low(grid, rc)
        basin = Set{CartesianIndex}()
        q = Vector{CartesianIndex}()
        push!(q, rc)
        while length(q) > 0
            c = popfirst!(q)
            push!(basin, c)
            for n ∈ neighbours(grid, c[1], c[2])
                (n ∉ basin && grid[n] != 9) && push!(q, n)
            end
        end
        return length(basin)
    end
    grid = _parse(input)
    sizes = [sizeof_basin_round_low(grid, low) for low ∈ find_lows(grid)]
    return prod(sort(sizes, rev=true)[1:3])
end

TEST = @aoc_splitlines("""
2199943210
3987894921
9856789892
8767896789
9899965678
""")

function samples()
    @assert part1(TEST) == 15
    @assert part2(TEST) == 1134
end

end # module AoC2021_09

aoc_main(@__FILE__, ARGS, 2021, 9)
